src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -74,13 +106,242 @@ public class ConsensusManager
+        /// </summary>
+        /// <remarks>
+        /// Requests that have too many blocks will be split in batches.
+        /// </remarks>
+        private void ProcessDownloadQueueLocked()
+        {
+            this.logger.LogTrace("()");
+
+            const long MaxUnconsumedBlocksDataBytes = 1048576 * 10; // 10 mb
+
+            while (this.toDownloadQueue.Count > 0)
+            {
+                BlockDownloadRequest request = this.toDownloadQueue.Peek();
+
+                long freeMb = MaxUnconsumedBlocksDataBytes - this.chainedHeaderTree.UnconsumedBlocksDataBytes;
+                this.logger.LogTrace("There is {0} MB available do download is.", freeMb);
+
+                if (freeMb <= 32)
+                    return;
+
+                long blocksToAsk = freeMb / this.blockPuller.AverageBlockSize;
+                this.logger.LogTrace("The slot of available blocks to ask is {0}.", blocksToAsk);
+
+                if (request.BlocksToDownload.Count < blocksToAsk)
+                {
+                    this.toDownloadQueue.Dequeue();
+                    this.blockPuller.RequestNewData(request);
+                }
+                else
+                {
+                    // split queue item in 2 pieces - one if size blocksToAsk and second is rest.Ask BP for first part, leave 2nd part in queue
+
+                    var newRequest = new BlockDownloadRequest()
+                    {
+                        DownloadedBlockDelegate = request.DownloadedBlockDelegate,
+                        BlocksToDownload = new List<uint256>()
+                    };
+
+                    for (int i = 0; i < blocksToAsk; i++)
+                    {
+                        uint256 hash = request.BlocksToDownload[0];
+                        request.BlocksToDownload.Remove(hash);
+                        newRequest.BlocksToDownload.Add(hash);
+                    }
+
+                    this.blockPuller.RequestNewData(newRequest);
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+    }
+
+    /// <summary>
+    /// A delegate that is used to send callbacks when a bock is downloaded from the of queued requests to downloading blocks. 
+    /// </summary>
+    /// <param name="blockPair">The pair of the block and its chained header.</param>
+    public delegate void ProcessDownloadedBlockDelegate(BlockPair blockPair);
+
+    /// <summary>
+    /// A request that holds information of blocks to download.
+    /// </summary>
+    public class BlockDownloadRequest
+    {
+        /// <summary>The list of block headers to download.</summary>
+        public List<uint256> BlocksToDownload { get; set; }
+
+        /// <summary>The delegate to send the downloaded blocks to.</summary>
+        public ProcessDownloadedBlockDelegate DownloadedBlockDelegate { get; set; }
```

Comment:
Maybe rename type and variable name to OnBlockDownloadedCallback?

Reply:
ok sure